setup.py

```diff
@@ -59,6 +59,90 @@ def dmf_data_files(root: str = DMF_DATA_ROOT) -> List[Tuple[str, List[str]]]:
     return file_list
 
 
+class ExtraDependencies:
+    """
+    A convenience shorthand to define and combine dependencies for ``extras_require``.
+
+    >>> extras = ExtraDependencies()
+    >>> extras["ui"]
+    ['requests', 'pint']
+    >>> list(extras)
+    ['ui', 'examples', 'dmf', 'surrogate', 'mlai', 'grid', 'optional']
+    >>> dict(extras)
+    {'ui': ['requests', 'pint'], 'examples': ['jupyter', 'nbconvert', 'nbformat'], ...}
+    """
+
+    ui = [
+        "requests",
+        "pint",
+    ]
+    examples = [
+        "jupyter",
+        "nbconvert",
+        "nbformat",
+    ]
+    dmf = (
+        [
+            # all modules relative to idaes.core.dmf
+            "jsonschema",  # commands, resource, workspace
+            "setuptools",  # provides pkg_resources?
+            "traitlets",  # dmfbase
+            "lxml",  # help
+            "seaborn",  # model_data (optional^2)
+            "PyPDF2",  # model_data (optional^2)
+            "colorama",  # util
+            "ipython",  # magics
+            "pyyaml",  # workspace
+            "tinydb",  # resourcedb
+            "xlrd",  # tables (implicitly by pandas.read_excel())
+            "openpyxl",  # tables (implicitly by pandas.read_excel())
+        ],
+    )
+    surrogate = [
+        "omlt==1.1",  # fix the version for now as package evolves
+    ]
+    mlai = [
+        "tensorflow",
```

Comment:
I agree this makes sense from a conceptual standpoint and that's how I originally grouped it. However, if we want to have one `extras_require` target to match the current set of dependencies, the "topical" targets must be such that `omlt` (which is installed by default now) and `tensorflow` (which is not) are separate.

More concretely, there's also the factor of `tensorflow` being of eating-into-my-ISP-data-caps size (not to mention the multiple widespread side effects caused by merely importing it), which IMO are all good reasons to limit as much as possible the circumstances in which it does get installed.

Note that the other "solution" for these constraints would be that *both* OMLT and Tensorflow are not part of the `complete` set of dependencies, which, as far as I'm concerned, I think we should consider as a possible choice ("installation backward-compatibility" considerations aside).

Reply:
We should check with the OMLT developers and see if it can be used without tensorflow or not. If there is a use case where you can do something without tensorflow then they can probably remain separate, but if you really need both then we should make OMLT optional as well.

Reply:
I've reworked the grouping a bit taking this comment into account. There's now an `omlt` target comprising `omlt` and `tensorflow`.